### PR TITLE
CI: Add checks for commit body ending format

### LINKED ISSUE #3026: [Feature Request] Check format of commit messages
# Feature Request

Without checking the ending of a commit message it can be improperly formatted like 7d1ba5661dfbc04c9f7c27da1c0144090109b17a which went past multiple reviews and passed tests, eventually merging onto main with the commit message:
```
Signed-off-by: Jonathan Woollett-Light <jcawl@amazon.co.uk>

shorter
```

## Describe the desired solution

Verify a commit message ends with `\nSigned-off-by: ...` (and there is only a single `Signed-off-by: ...`) e.g.:
```
some message

Signed-off-by: Jonathan Woollett-Light <jcawl@amazon.co.uk>
```

Followed by 0 or more `\nCo-authored-by: ...`s (and none precede `Signed-off-by: ...`) e.g.:
```
some message

Signed-off-by: Jonathan Woollett-Light <jcawl@amazon.co.uk>
Co-authored-by: Alexandru Cihodaru <cihodar@amazon.com>
```

## Checks

- [x] Have you searched the Firecracker Issues database for similar requests?
- [x] Have you read all the existing relevant Firecracker documentation?
- [x] Have you read and understood Firecracker's core tenets?

### PR DESCRIPTION
**Pull request by CuriousCorrelation**
# Reason for This PR
Add checks for commit body ending format, closes #3026 

## Description of Changes

Adds test for commit body ending format
that makes sure body ends with 'Signed-off-by' or 'Co-authored-by'
and no other trailing line

## Examples

``` sh
   1   │ A commit message
   2   │
   3   │ some message
   4   │ some other message
   5   │ Signed-off-by: someone@somewhere.com
```

``` sh
❯ cat gitlint_example_1 | gitlint -C .gitlint --extra-path tests/framework/gitlint_rules.py
```

-----------

``` sh
   1   │ A commit message
```

``` sh
❯ cat gitlint_example_2 | gitlint -C .gitlint --extra-path tests/framework/gitlint_rules.py
1: UC2 Body does not contain a 'Signed-off-by' line
3: B6 Body message is missing
```

-----------

``` sh
   1   │ A commit message
   2   │
   3   │ some other message
   4   │ Signed-off-by: someone@somewhere.com
   5   │ Co-authored-by: someone_else@somewhere_else.com
```

``` sh
❯ cat gitlint_example_3 | gitlint -C .gitlint --extra-path tests/framework/gitlint_rules.py
```
-----------

``` sh
   1   │ A commit message
   2   │
   3   │ Signed-off-by: someone@somewhere.com
   4   │ Co-authored-by: someone_else@somewhere_else.com
   5   │ some other message
```

``` sh
❯ cat gitlint_example_4 | gitlint -C .gitlint --extra-path tests/framework/gitlint_rules.py
2: UC2 Last line in commit message body should be 'Signed-off-by'
```
-----------

``` sh
   1   │ A commit message
   2   │
   3   │ Signed-off-by: someone@somewhere.com
   4   │ Co-authored-by: someone_else@somewhere_else.com
   5   │
   6   │ some other message
```

``` sh
❯ cat gitlint_example_5 | gitlint -C .gitlint --extra-path tests/framework/gitlint_rules.py
3: UC2 Last line in commit message body should be 'Signed-off-by'
```

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] All added/changed functionality is tested.

### CODE REVIEW

**Comment by JonathanWoollett-Light** [comment]
> UC2 Last line in commit message body should be 'Signed-off-by'

This is not accurate the last line should be either `Signed-off-by` or `Co-authored-by`.

---

To summarize a few points a cleaner implementation may look like:
```python
def validate(self, commit):
    """Validate user defined gitlint rules."""

    # Utilities
    rtn = lambda str,i: [RuleViolation(id,str,None,i)]
    CO_AUTH = "Co-authored-by:"
    SIG = "Signed-off-by:"
    
    iter = enumerate(commit.message.body)
    
    # Checks commit message contains a `SIG` string
    found = False
    for i, line in iter:
        # We check that no co-authors are declared before signatures.
        if line.startswith(CO_AUTH): return rtn(f"'{CO_AUTH}' found before '{SIG}'",i)
        elif line.startswith(SIG):
            found = True
            break
    # If no signature was found in the message (before `iter` ended)
    if not found: return rtn(f"'{SIG}' not found in commit message", None)

    # Checks lines following signature are either signatures or co-authors
    for i, line in iter:
        if line.startswith(SIG): continue
        # Once we encounter the first co-author, we no longer accept signatures
        elif line.startswith(CO_AUTH): break
        else: return rtn(
            f"Non '{CO_AUTH}' or '{SIG}' string found following 1st '{SIG}'", i
        )

    # Checks lines following co-author are only additional co-authors.
    for i, line in iter:
        if not line.startswith(CO_AUTH):
            return rtn(f"Non '{CO_AUTH}' string found after 1st '{CO_AUTH}'", i)

    # Return no errors
    return []
```
To argue a few points:
- We can abstract some complexity away using a lambda `rtn` which is subjective but something I prefer over repeating formatting.
- As we re-use the strings `"Co-authored-by:"` and `"Signed-off-by:"` it would be cleaner to use local constants.
- We can create 1 iterator with `iter = enumerate(commit.message.body)` which avoiding both needing to call `enumerate` twice and to offset set this like `enumerate(message_lines[index:])`.
- I find generally a flatter structure is more readable, by using `found` in the first loop we avoid nesting the rest of the code. This also allows us to more clearly segment the sections.
- I think `i` is a more concise identifier than `index` in this context, but this is subjective.
- When combining the checks with `after_line.startswith("Signed-off-by") or after_line.startswith("Co-authored-by") ...` you loose information. It can be discouraging to use multiple if statements when you need a newline for every one is it may begin to appear messy, here I think inline if statement is fine.
- I think as a general rule it is best to think about how to provide the maximum information and avoid discarding information by joining multiple cases for convenience.

What are your thoughts?

**Comment by CuriousCorrelation** [comment]
Makes perfect sense. Thank you for taking time to explain your thought process.

I was initially going to constrain 'Signed-off-by' to be before 'Co-authored-by' like in your impl
``` sh
   1   │ A commit message
   2   │
   3   │ some other message
   4   │ Co-authored-by: someone_else@somewhere_else.com
   5   │ Signed-off-by: someone@somewhere.com
```
``` sh
2: UC2 'Co-authored-by:' found before 'Signed-off-by:'
```
but after looking at some samples including for linux kernel [[1](https://github.com/torvalds/linux/commit/30ffa74a04b6e48080f3181b29d2693b8486ca1e), [2](https://github.com/torvalds/linux/commit/3743bec6120ae0748cb3fda6ff80a690117ef1f3), [3](https://github.com/torvalds/linux/commit/dda5384313a40ecbaafd8a9a80f47483255e4c4d)] I couldn't to find a common standard/consensus.

Although I agree more with this impl. I believe setting a standard upfront would lead to a lot less confusion/debate in the future.

I think
``` python
    rtn = lambda str,i: [RuleViolation(id,str,None,i)]
```
should be
``` python
    rtn = lambda str,i: [RuleViolation(self.id,str,None,i)]
```

and error message on 

``` sh
   1   │ A commit message
```

says 

``` sh
-: UC2 'Signed-off-by:' not found in commit message
```

perhaps "not found in commit message body" would be more clear to someone fixing it as 

``` sh
git commit -m "A commit message Signed-off-by: some-email"
```

instead of
`git commt` or 
``` sh
git commit -m "A commit message" -m "Signed-off-by: some-email"
```

--- 

I have edited file to reflect your code so let me know if I should make these changes and commit. Also if I should link to your comment in my commit body.

**Comment by JonathanWoollett-Light** [comment]
> but after looking at some samples including for linux kernel [[1](https://github.com/torvalds/linux/commit/30ffa74a04b6e48080f3181b29d2693b8486ca1e), [2](https://github.com/torvalds/linux/commit/3743bec6120ae0748cb3fda6ff80a690117ef1f3), [3](https://github.com/torvalds/linux/commit/dda5384313a40ecbaafd8a9a80f47483255e4c4d)] I couldn't to find a common standard/consensus.

Here I have to admit I'm really only going off what works with GitHub (having Co-Authors preceding the signatures does not seem to work). If you make a PR with it the other way around you will need to adjust it, and thus I think we should test for this.

> I believe setting a standard upfront would lead to a lot less confusion/debate in the future.

If you think this test may not illustrate a standard clear enough it may be worth adding an example under [CONTRIBUTING.md#contribution-quality-standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards) e.g. 
> A good commit message may look like
> ```
> A descriptive title under <x> characters
>
> A concise description where each line
> is under <y> characters.
>
> Signed-Off-By <A full name> <A email>
> Co-Authored-By: <B full name> <B email>
> ```
(replace `x` and `y` with the current values used for tests)

For both:
>     rtn = lambda str,i: [RuleViolation(self.id,str,None,i)]
and:
> perhaps "not found in commit message body" would be more clear to someone fixing it as

I agree these are better.

**Comment by CuriousCorrelation** [comment]
> If you think this test may not illustrate a standard clear enough

No I think 
``` sh
2: UC2 'Co-authored-by:' found before 'Signed-off-by:'
```
should be clear enough indication. By "I agree more with this impl. I believe setting a..." I meant what you suggest seems better than having unclear standards.

Although this 

> example under [CONTRIBUTING.md#contribution-quality-standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards) 
>  A good commit message may look like
>  ```
>  A descriptive title under <x> characters
>  
>  A concise description where each line
>  is under <y> characters.
>  
>  Signed-Off-By <A full name> <A email>
>  Co-Authored-By: <B full name> <B email>
>  ```

seems like an excellent idea I'd be happy to add this and create a new PR!

**Comment by JonathanWoollett-Light** [comment]
> seems like an excellent idea. I'd be happy to add this and create a new PR!

I think this fits under this PR as it specifically relates to the commit messages (which I think is a narrow enough topic). You could create a new PR but I think it would just be easier to add a commit.

**Comment by JonathanWoollett-Light** [comment]
I also think the description of the class is a little inaccurate:
```
Make sure that each commit ends with a signature.
Either ending with a "Signed-off-by" line
or "Signed-off-by" followed by "Co-authored-by".
```
I might suggest something like:
```
Checks each commit ends with a 1 or 
more signatures ("Signed-off-by") followed by 0 
or more co-authors ("Co-authored-by").
```

**Comment by JonathanWoollett-Light** [comment]
Just noticed rather than:
```
A descriptive title under 50 characters

A concise description where each line is under 72 characters.
```
it should be:
```
A descriptive title of 50 characters or less

A concise description where each line is 72 characters or less.
```
As [.gitlint](https://github.com/firecracker-microvm/firecracker/blob/main/.gitlint) specifies the max allowed as `50` and `72`.

**Comment by CuriousCorrelation** [comment]
Done!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -68,6 +68,15 @@ testing. Your contribution needs to meet the following standards:
 - Document all your public functions.
 - Add a descriptive message for each commit. Follow
   [commit message best practices](https://github.com/erlang/otp/wiki/writing-good-commit-messages).
+- A good commit message may look like
+  ```
+  A descriptive title of 50 characters or less
+  
+  A concise description where each line is 72 characters or less.
```
Comment (gregorybrzeski) [inline CONTRIBUTING.md:75]: I would suggest to use word "fewer" instead of "less".

**Comment by JonathanWoollett-Light** [comment]
Nice!

Please rebase onto `main` and check all tests pass by running `./tools/devtool test`.

**Comment by CuriousCorrelation** [comment]
These are the test results 

About `gitlint` on recent commits
```
_________ test_gitlint ___________
integration_tests/style/test_gitlint.py:20: in test_gitlint
    utils.run_cmd('gitlint --commits origin/main..HEAD'
framework/utils.py:567: in run_cmd
    return run_cmd_sync(cmd=cmd,
framework/utils.py:443: in run_cmd_sync
    raise ChildProcessError(output_message)
E   ChildProcessError:
E   [2860] Command:
E   gitlint --commits origin/main..HEAD -C ../.gitlint --extra-path framework/gitlint_rules.py
E   [2860] stdout:
E
E   [2860] stderr:
E   Commit 0afd89468d:
E   2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E   Commit f2a017d364:
E   2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E   Commit 2fc473bbc4:
E   2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E   Returned error code: 3
E   stderr:
E   Commit 0afd89468d:
E   2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E   Commit f2a017d364:
E   2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E   Commit 2fc473bbc4:
E   2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'

During handling of the above exception, another exception occurred:
framework/scheduler.py:132: in call_wrapper
    original_function(*args, **kwargs))
integration_tests/style/test_gitlint.py:24: in test_gitlint
    assert False, "Commit message violates gitlint rules: {}".format(error)
E   AssertionError: Commit message violates gitlint rules:
E     [2860] Command:
E     gitlint --commits origin/main..HEAD -C ../.gitlint --extra-path framework/gitlint_rules.py
E     [2860] stdout:
E
E     [2860] stderr:
E     Commit 0afd89468d:
E     2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E     Commit f2a017d364:
E     2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E     Commit 2fc473bbc4:
E     2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E     Returned error code: 3
E     stderr:
E     Commit 0afd89468d:
E     2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E     Commit f2a017d364:
E     2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E     Commit 2fc473bbc4:
E     2: UC2 Non 'Co-authored-by:' or 'Signed-off-by:' string found following 1st 'Signed-off-by:'
E
E   assert False
__________________
```

About `CONTRIBUTING`

```
E   [2923] stdout:
E   .././CONTRIBUTING.md:74: MD009 Trailing spaces
E   .././CONTRIBUTING.md:76: MD009 Trailing spaces
E   .././CONTRIBUTING.md:72: MD031 Fenced code blocks should be surrounded by blank lines
E   .././CONTRIBUTING.md:79: MD031 Fenced code blocks should be surrounded by blank lines
```

About `gitlint_rules` file

```
E   ************* Module gitlint_rules
E   framework/gitlint_rules.py:31:8: W0622: Redefining built-in 'iter' (redefined-builtin)
E   framework/gitlint_rules.py:28:8: C0103: Variable name "CO_AUTH" doesn't conform to '[a-z_][a-z0-9_]{1,30}$' pattern (invalid-name)
E   framework/gitlint_rules.py:29:8: C0103: Variable name "SIG" doesn't conform to '[a-z_][a-z0-9_]{1,30}$' pattern (invalid-name)
E   framework/gitlint_rules.py:37:12: R1705: Unnecessary "elif" after "return", remove the leading "el" from "elif" (no-else-return)
E   framework/gitlint_rules.py:49:12: R1724: Unnecessary "elif" after "continue", remove the leading "el" from "elif" (no-else-continue)
E
E   Returned error code: 28
```

Beside these, all others were platform related like

```
ERROR integration_tests/performance/test_block_performance.py - FileNotFoundError: [Errno 2] No such file or directory: '/firecracker/tests/integration_tests/performance/configs/test_block_performance_config_5.15.json'
```

5 failed, 206 passed, 4 skipped, 4 deselected

File: tests/framework/gitlint_rules.py [hunk 2]
```diff
@@ -8,20 +8,70 @@
...
 
-class SignedOffBy(CommitRule):
-    """Make sure that each commit contains a "Signed-off-by" line."""
+class EndsSigned(CommitRule):
+    """Checks commit message body formatting.
+
+    Makes sure each commit message body ends with
+    1 or more signatures ("Signed-off-by"), followed by
+    0 or more co-authors ("Co-authored-by").
+    """
```
Comment (CuriousCorrelation) [inline tests/framework/gitlint_rules.py:17]: All tests pass now (besides platform specific). There were several changes to the original code. I have selected some that seem especially important.

File: tests/framework/gitlint_rules.py [hunk 3]
```diff
@@ -8,20 +8,70 @@
...
     id = "UC2"
 
     def validate(self, commit):
         """Validate user defined gitlint rules."""
-        for line in commit.message.body:
-            if line.startswith("Signed-off-by"):
-                return []
+        # Utilities
+        def rtn(stmt, i):
+            return [RuleViolation(self.id, stmt, None, i)]
```
Comment (CuriousCorrelation) [inline tests/framework/gitlint_rules.py:29]: This was a lambda before, now `def`

E731 do not assign a lambda expression, use a def

File: tests/framework/gitlint_rules.py [hunk 4]
```diff
@@ -8,20 +8,70 @@
...
+
+        co_auth = "Co-authored-by:"
+        sig = "Signed-off-by:"
```
Comment (CuriousCorrelation) [inline tests/framework/gitlint_rules.py:32]: Capitalized variables before, now small cased

File: tests/framework/gitlint_rules.py [hunk 5]
```diff
@@ -8,20 +8,70 @@
...
+            # Once we encounter the first co-author,
+            # we no longer accept signatures
+            if line.startswith(co_auth):
+                break
+
+            return rtn(
+                (f"Non '{co_auth}' or '{sig}' string found "
+                 f"following 1st '{sig}'"),
+                i,
+            )
```
Comment (CuriousCorrelation) [inline tests/framework/gitlint_rules.py:66]: Broken into two line to conform to 80 char limit

File: tests/framework/gitlint_rules.py [hunk 6]
```diff
@@ -8,20 +8,70 @@
...
+        # If no signature was found in the message
+        # (before `message_iter` ended)
+        if not found:
+            return rtn(f"'{sig}' not found in commit message body", None)
+
+        # Checks lines following signature are
+        # either signatures or co-authors
+        for i, line in message_iter:
+            if line.startswith(sig) or not line.strip():
+                continue
```
Comment (CuriousCorrelation) [inline tests/framework/gitlint_rules.py:55]: This fixes false positives on `''` (blanks)

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -68,6 +68,17 @@ testing. Your contribution needs to meet the following standards:
...
 - Add a descriptive message for each commit. Follow
   [commit message best practices](https://github.com/erlang/otp/wiki/writing-good-commit-messages).
+- A good commit message may look like
+
+  ```
+  A descriptive title of 50 characters or fewer
+
+  A concise description where each line is 72 characters or fewer.
+
+  Signed-Off-By <A full name> <A email>
```
Comment (CuriousCorrelation) [inline CONTRIBUTING.md:78]: MD009 and MD031 are fixed now. Didn't need to rearrange.

**Comment by CuriousCorrelation** [comment]
@JonathanWoollett-Light Latest commit is up, all tests are green! Let me know if I missed something.

**Review by JonathanWoollett-Light — DISMISSED** [review]

**Comment by JonathanWoollett-Light** [comment]
Please rebase onto main.

**Comment by CuriousCorrelation** [comment]
@JonathanWoollett-Light Done, tests are green

**Review by pb8o — APPROVED** [review]

**Review by JonathanWoollett-Light — APPROVED** [review]

**Comment by JonathanWoollett-Light** [comment]
@CuriousCorrelation Thank you for your contribution

**Comment by CuriousCorrelation** [comment]
@JonathanWoollett-Light It was an awesome experience, got to learn a lot about the project and testing/CI systems in general. 
All thanks to your patience and help along the way! I truly appreciate it.

**Pull request merged**